### PR TITLE
Added strict comparisons and some type annotations

### PR DESCRIPTION
**Pull request by Nyholm**
TODO: 
- [ ] Validate the type annotations with the specification.

### CODE REVIEW

File: src/Request.php [hunk 1]
```diff
@@ -66,10 +66,10 @@ public function getRequestTarget(): string
         }
 
         $target = $this->uri->getPath();
-        if ('' == $target) {
+        if ('' === $target) {
             $target = '/';
         }
-        if ('' != $this->uri->getQuery()) {
+        if ('' !== $this->uri->getQuery()) {
```
Comment (sagikazarmark) [inline src/Request.php]: Why Yoda in an expression like this? (Just asking, not looking for a flame war, though IMO this is too much consistency)

File: src/Uri.php [hunk 2]
```diff
@@ -143,7 +143,7 @@ public function withScheme($scheme): self
     public function withUserInfo($user, $password = null): self
     {
         $info = $user;
-        if ('' != $password) {
+        if ('' !== $password) {
```
Comment (sagikazarmark) [inline src/Uri.php]: Password can be null

File: src/Response.php [hunk 3]
```diff
@@ -137,7 +137,7 @@ public function withStatus($code, $reasonPhrase = ''): self
 
         $new = clone $this;
         $new->statusCode = (int) $code;
-        if ('' == $reasonPhrase && isset(self::$phrases[$new->statusCode])) {
+        if ('' === $reasonPhrase && isset(self::$phrases[$new->statusCode])) {
```
Comment (sagikazarmark) [inline src/Response.php]: Nothing protects from passing a null reason phrase here which would bypass this condition

File: src/Uri.php (hunk 2 above)
Comment (Zegnat) [reply src/Uri.php]: Good catch.

Something that I didn’t even think about: is empty string (`''`) a valid HTTP password? I.e. is `https://zegnat:@example.com/` different from `https://zegnat@example.com/`? If an empty string is valid, this if-clause should specifically check `null !== $password`.

Comment (sagikazarmark) [reply src/Uri.php]: I think it is valid, but I'm not sure it's any different from an no password at all. One could check the RFCs, but my gut tells that implementations might not always follow the RFC here.

Comment (Zegnat) [reply src/Uri.php]: A quick look at [RFC 3986](https://tools.ietf.org/html/rfc3986), as cited by PSR-7, tells me that an empty string indicates no password. So really, `''` equals `null` as far as URIs are concerned. This also matches with [the WHATWG URL Standard](https://url.spec.whatwg.org/) (that tries to bring several other standards together), where [password is an empty string by default](https://url.spec.whatwg.org/#concept-url-password).

Part of me wants to argue that the method signature is wrong, and should have been `withUserInfo($user, $password = '')`.

**Comment by Nyholm** [comment]
Thank you for the reviews. I've made some updates.

**Pull request merged**